### DISCUSSION TITLE
Format Email Timestamp

### DISCUSSION THREAD

**Discussion by tugatobito** [jertel/elastalert2#121] [Q&A]
Is there a way to format the timestamp that goes on the alert?

Im using the default timestamp field.

![image](https://user-images.githubusercontent.com/83758798/117276452-d2ae7080-ae56-11eb-8610-e432e3f56a85.png)
I would like something "6 May of 2021 at 9:30am".

I've being trying to change the python util.py file but without success, the timefield always goes to back to that mess.

**Comment by jertel** [comment]
If you're referring to the list of match key/value pairs that get dumped into the alert body then no, I don't believe you can format the values. It's using the default python string conversion `value_str = str(value)` found in `alerts.py::_add_match_items::127`.

**Reply by tugatobito** [reply]
And theres literally no way to change this? Even in the code itself?

It just seems odd that you can't customize the timestamp of the alert.

**Reply by jertel** [reply]
You can do almost anything if you're up for changing the code. My response was meant to convey that you can't format the value via a change to your rule_xyz.yaml file. If you are up to some Python coding then yes you can change the format of those key/value pairs. You could use an if or switch block to detect the type of the value, and if it's a timestamp then call a format function on it. You would make that change around the line I pasted in my first response. Can you paste more of your alert email so I can be sure we're talking about the same part of that email body?

**Reply by tugatobito** [reply]
It's just a simple email alert, the timestamp goes into the body.

![image](https://user-images.githubusercontent.com/83758798/117294147-38f0be80-ae6a-11eb-9aa2-6b209c211db5.png)

I've been trying to change the code, but without success... my knowledge of Python might not be enough for it.

**Reply by jertel** [reply]
Ok, in that case you will need to make changes on a different line. Look at alerts.py around line 83.

```
                    if alert_value:
                        alert_text_values[i] = alert_value
```

You will need to check the type of `alert_value` and if it's a timestamp then format it.

**Reply by tugatobito** [reply]
I found the line and tried with this.
![image](https://user-images.githubusercontent.com/83758798/117296648-3348a800-ae6d-11eb-89e8-e2da5d27f8a4.png)

or even

![image](https://user-images.githubusercontent.com/83758798/117297285-f03b0480-ae6d-11eb-907b-e3817b882b35.png)

but the output is the same. It seems that the code just doesnt change or something.

**Reply by jertel** [reply]
Sounds like you're not modifying the code that's actually being executed. I can't offer much assistance since I am not familiar with your setup.

**Reply by nsano-rururu** [reply]
@tugatobito 

I feel that the settings of timestamp_field and timestamp_type and timestamp_format are related.

https://elastalert2.readthedocs.io/en/latest/ruletypes.html#timestamp-type

https://github.com/Yelp/elastalert/issues/616
https://github.com/Yelp/elastalert/issues/1701
>timestamp_field: timestamp
>timestamp_type: custom
>timestamp_format: "%d.%m.%Y %H:%M:%S"
https://github.com/Yelp/elastalert/issues/996
>timestamp_field: timestamp
>timestamp_type: custom
>timestamp_format: '%Y-%m-%d %H:%M:%S.%f'
>timestamp_format_expr: 'ts[:23] + ts[26:]'

**Reply by nsano-rururu** [reply]
https://github.com/jertel/elastalert2/blob/master/elastalert/loaders.py#L287
```
        rule.setdefault('timestamp_field', '@timestamp')
        rule.setdefault('filter', [])
        rule.setdefault('timestamp_type', 'iso')
        rule.setdefault('timestamp_format', '%Y-%m-%dT%H:%M:%SZ')
```
https://github.com/jertel/elastalert2/blob/master/elastalert/loaders.py#L308
```
        elif rule['timestamp_type'] == 'custom':
            def _ts_to_dt_with_format(ts):
                return ts_to_dt_with_format(ts, ts_format=rule['timestamp_format'])

            def _dt_to_ts_with_format(dt):
                ts = dt_to_ts_with_format(dt, ts_format=rule['timestamp_format'])
                if 'timestamp_format_expr' in rule:
                    # eval expression passing 'ts' and 'dt'
                    return eval(rule['timestamp_format_expr'], {'ts': ts, 'dt': dt})
                else:
                    return ts

            rule['ts_to_dt'] = _ts_to_dt_with_format
            rule['dt_to_ts'] = _dt_to_ts_with_format
```
